### PR TITLE
Refactor backoffs + retries

### PR DESCRIPTION
**Pull request by Horneth**
This incorporates changes coming from discussions in https://github.com/broadinstitute/cromwell/pull/389 + solves https://broadinstitute.atlassian.net/browse/DSDEEPB-2644
It also adds jiggling to `TryUtil.withRetry` which now uses Google implementation of exponential backoff.

### CODE REVIEW

**Comment by Horneth** [comment]
also @kshakir

File: src/main/scala/cromwell/util/Backoff.scala [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import com.google.api.client.util.ExponentialBackOff
+
+import scala.concurrent.duration.{Duration, FiniteDuration}
+
+sealed trait Backoff {
+  def backoffMillis: Long
+  def next: Backoff
+}
+
+final class InitialGapBackoff(initialGapMillis: FiniteDuration, googleBackoff: ExponentialBackOff) extends Backoff {
```
Comment (mcovarr) [inline src/main/scala/cromwell/util/Backoff.scala]: any reason these aren't `case` classes?

File: src/main/scala/cromwell/util/Backoff.scala [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      .setMultiplier(multiplier)
+      .setMaxElapsedTimeMillis(Int.MaxValue)
+      .build())
+  }
+
+  override val backoffMillis = initialGapMillis.toMillis
+  override def next = new SimpleBackoff(googleBackoff)
+}
+
+final class SimpleBackoff(googleBackoff: ExponentialBackOff) extends Backoff {
```
Comment (mcovarr) [inline src/main/scala/cromwell/util/Backoff.scala]: Maybe the name should have something about it being an exponential backoff.

File: src/main/scala/cromwell/util/Backoff.scala [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+      .setInitialIntervalMillis(initialInterval.toMillis.toInt)
+      .setMaxIntervalMillis(maxInterval.toMillis.toInt)
+      .setMultiplier(multiplier)
+      .setMaxElapsedTimeMillis(Int.MaxValue)
+      .build())
+  }
+
+  override def backoffMillis = googleBackoff.nextBackOffMillis()
+  override def next = this
+}
```
Comment (mcovarr) [inline src/main/scala/cromwell/util/Backoff.scala]: A little bit of Scaladoc on each of them would be nice

**Comment by mcovarr** [comment]
:+1:

**Comment by mcovarr** [comment]
Quoth the wheel: @cjllanwarne

File: src/main/scala/cromwell/util/Backoff.scala (hunk 1 above)
Comment (Horneth) [inline src/main/scala/cromwell/util/Backoff.scala]: I wasn't able to define a second constructor with a case class - IntelliJ was whining.

Comment (geoffjentry) [inline src/main/scala/cromwell/util/Backoff.scala]: It should probably be an apply anyways

Comment (kshakir) [inline src/main/scala/cromwell/util/Backoff.scala]: +1 on trying an apply with a case class

File: src/main/scala/cromwell/util/TryUtil.scala [hunk 4]
```diff
@@ -69,19 +76,24 @@ object TryUtil {
...
+        val retryCountMessage = {
+          val transientMessage = if (transient) " - This error is transient and does not count against the retry limit." else ""
+          if (retryLimit.getOrElse(0) > 0 || transient) s" (${newLimit.getOrElse(0)} more retries)$transientMessage" else ""
+        }
+        val pollingInterval = backoff.backoffMillis
         val retryMessage = s"Retrying in $pollingInterval$retryCountMessage..."
         failMessage foreach { m => logger.warn(s"$m.  $retryMessage") }
 
-        Thread.sleep(pollingInterval.toMillis)
+        Thread.sleep(pollingInterval)
```
Comment (kshakir) [inline src/main/scala/cromwell/util/TryUtil.scala]: I never paid attention to this before. This "smells" funny, but I'm not looking for any changes right now. I feel like to be a better async citizen, this logic should be using more akka, and scheduling some message to be sent again to some actor.

**Comment by kshakir** [comment]
:+1: after other comments addressed. @cjllanwarne is still welcome to review also if he likes.

File: src/main/scala/cromwell/util/TryUtil.scala (hunk 4 above)
Comment (geoffjentry) [inline src/main/scala/cromwell/util/TryUtil.scala]: There's actually a ticket to remove this, it just isn't happened yet

**Pull request merged**